File: onecard-server/src/modules/game/services/game-ai.service.ts
import {
  Injectable,
  Logger,
  BadRequestException,
  NotFoundException,
} from '@nestjs/common';
import type { GameState } from '@/modules/game/domain/types/gameState';
import type { GameAction } from '@/modules/game/domain/types/gameAction';
import type { PokerCardPropsWithId } from '@/modules/game/domain/types/pokerCard';
import type { Player } from '@/modules/game/domain/types/gamePlayer';
import { findPlayableCardBruteForce } from '@/modules/game/domain/state/gamePlayers';
import { GameEngineService } from '@/modules/game/services/game-engine.service';
import {
  applySpecialEffectAction,
  drawCardAction,
  nextTurnAction,
  playCardAction,
  EngineStepResult,
} from '@/modules/game/domain/engine/gameEngine';
import { isValidPlay } from '@/modules/game/domain/utils/cardUtils';
import { OnnxPolicyService } from '@/modules/game/inference/onnx-policy.service';

interface AiTurnResult {
  state: GameState;
  actions: GameAction[];
  result: EngineStepResult;
}

interface AiTurnContext {
  gameId?: string;
}

@Injectable()
export class GameAiService {
  private readonly logger = new Logger(GameAiService.name);

  public constructor(
    private readonly gameEngine: GameEngineService,
    private readonly onnxPolicyService: OnnxPolicyService,
  ) {}

  public async playWhileAiTurn(
    state: GameState,
    context?: AiTurnContext,
  ): Promise<EngineStepResult | null> {
    if (state.settings.mode !== 'single' || !this.isAiTurn(state)) {
      return null;
    }

    if (state.settings.difficulty === 'medium') {
      return this.playWithOnnx(state, context);
    }

    const turnResult = this.executeTurn(state, context);
    if (!turnResult) {
      return null;
    }

    return {
      state: turnResult.state,
      done: turnResult.state.gameStatus === 'finished',
      info: {
        aiActions: turnResult.actions,
      },
    };
  }

  private executeTurn(
    state: GameState,
    context?: AiTurnContext,
  ): AiTurnResult | null {
    const player = state.players.at(state.currentPlayerIndex);
    if (!player?.isAI) {
      return null;
    }

    let currentState = state;
    let lastResult: EngineStepResult | undefined;
    const actions: GameAction[] = [];

    const topCard = currentState.discardPile[0];
    const cardToPlayResult = findPlayableCardBruteForce(
      player.hand,
      topCard,
      currentState.damage,
    );

    if (cardToPlayResult) {
      const playableIndex = player.hand.findIndex(
        (c) => c.id === cardToPlayResult.id,
      );
      const playOutcome = this.applyAction(
        currentState,
        playCardAction(currentState.currentPlayerIndex, playableIndex),
        context,
      );
      currentState = playOutcome.state;
      lastResult = playOutcome.result;
      this.pushAction(actions, playOutcome.result);

      if (this.hasSpecialEffect(cardToPlayResult)) {
        const effectOutcome = this.applyAction(
          currentState,
          applySpecialEffectAction(cardToPlayResult),
          context,
        );
        currentState = effectOutcome.state;
        lastResult = effectOutcome.result;
        this.pushAction(actions, effectOutcome.result);
      }
    } else {
      const drawOutcome = this.applyAction(
        currentState,
        drawCardAction(Math.max(1, currentState.damage)),
        context,
      );
      currentState = drawOutcome.state;
      lastResult = drawOutcome.result;
      this.pushAction(actions, drawOutcome.result);

      const refreshedPlayer = currentState.players.at(
        currentState.currentPlayerIndex,
      );
      if (refreshedPlayer) {
        const refreshedTopCard = currentState.discardPile[0];
        const refreshedCardToPlayResult = findPlayableCardBruteForce(
          refreshedPlayer.hand,
          refreshedTopCard,
          currentState.damage,
        );

        if (refreshedCardToPlayResult) {
          const playableIndex = refreshedPlayer.hand.findIndex(
            (c) => c.id === refreshedCardToPlayResult.id,
          );
          const playOutcome = this.applyAction(
            currentState,
            playCardAction(currentState.currentPlayerIndex, playableIndex),
            context,
          );
          currentState = playOutcome.state;
          lastResult = playOutcome.result;
          this.pushAction(actions, playOutcome.result);

          if (this.hasSpecialEffect(refreshedCardToPlayResult)) {
            const effectOutcome = this.applyAction(
              currentState,
              applySpecialEffectAction(refreshedCardToPlayResult),
              context,
            );
            currentState = effectOutcome.state;
            lastResult = effectOutcome.result;
            this.pushAction(actions, effectOutcome.result);
          }
        }
      }
    }

    if (currentState.gameStatus !== 'finished') {
      const nextOutcome = this.applyAction(
        currentState,
        nextTurnAction(),
        context,
      );
      currentState = nextOutcome.state;
      lastResult = nextOutcome.result;
      this.pushAction(actions, nextOutcome.result);
    }

    return {
      state: currentState,
      actions,
      result: lastResult,
    };
  }

  private applyAction(
    state: GameState,
    action: GameAction,
    context?: AiTurnContext,
  ): { state: GameState; result: EngineStepResult } {
    this.logAiAction(
      action,
      state.players.at(state.currentPlayerIndex),
      context,
    );
    const result = this.gameEngine.step(state, action);
    return {
      state: result.state,
      result,
    };
  }

  private pushAction(actions: GameAction[], result: EngineStepResult): void {
    if (result.info?.action) {
      actions.push(result.info.action as GameAction);
    }
  }

  private logAiAction(
    action: GameAction,
    actor: Player | undefined,
    context?: AiTurnContext,
  ): void {
    if (!actor?.isAI) {
      return;
    }
    const metadata = {
      event: 'ai-action',
      gameId: context?.gameId ?? null,
      playerId: actor.id,
      playerName: actor.name,
      actionType: action.type,
      payload: 'payload' in action ? action.payload : null,
    };
    this.logger.log(`[AI] ${JSON.stringify(metadata)}`);
  }

  private findPlayableCardIndex(
    hand: PokerCardPropsWithId[],
    topCard: PokerCardPropsWithId | undefined,
    damage: number,
  ): number {
    if (!topCard) {
      return -1;
    }
    return hand.findIndex((card) => isValidPlay(card, topCard, damage));
  }

  private hasSpecialEffect(card: PokerCardPropsWithId): boolean {
    if (card.isJoker) {
      return true;
    }
    const specialRanks = new Set([1, 2, 11, 12, 13]);
    return card.rank !== undefined && specialRanks.has(card.rank);
  }

  private async playWithOnnx(
    state: GameState,
    context?: AiTurnContext,
  ): Promise<EngineStepResult | null> {
    try {
      let currentState = state;
      const actions: GameAction[] = [];
      let lastResult: EngineStepResult | undefined;

      const prediction =
        await this.onnxPolicyService.predictAction(currentState);
      const { payload, actionIndex } = prediction;
      const isDraw = payload.type === 'DRAW_CARD';
      const amount = isDraw ? (payload.amount ?? 1) : 1;
      const playerIndex = isDraw ? 0 : (payload.playerIndex ?? 0);
      const cardIndex = isDraw ? 0 : (payload.cardIndex ?? 0);

      const firstAction: GameAction = isDraw
        ? drawCardAction(amount)
        : playCardAction(playerIndex, cardIndex);

      this.logger.log(
        `[AI][ONNX] actionIndex=${String(actionIndex)} payload=${JSON.stringify(payload)}`,
      );

      const firstOutcome = this.applyAction(currentState, firstAction, context);
      currentState = firstOutcome.state;
      lastResult = firstOutcome.result;
      this.pushAction(actions, firstOutcome.result);

      if (!isDraw) {
        const player = state.players[state.currentPlayerIndex];
        const cardToPlay = player.hand[cardIndex];
        if (cardToPlay && this.hasSpecialEffect(cardToPlay)) {
          const effectOutcome = this.applyAction(
            currentState,
            applySpecialEffectAction(cardToPlay),
            context,
          );
          currentState = effectOutcome.state;
          lastResult = effectOutcome.result;
          this.pushAction(actions, effectOutcome.result);
        }
      }

      if (isDraw && currentState.gameStatus !== 'finished') {
        try {
          const secondPrediction =
            await this.onnxPolicyService.predictAction(currentState);
          const { payload: p2 } = secondPrediction;

          if (p2.type === 'PLAY_CARD') {
            const p2PlayerIdx = p2.playerIndex ?? 0;
            const p2CardIdx = p2.cardIndex ?? 0;

            const p2Player =
              currentState.players[currentState.currentPlayerIndex];
            if (p2Player && p2Player.hand[p2CardIdx]) {
              const secondAction = playCardAction(p2PlayerIdx, p2CardIdx);
              const secondOutcome = this.applyAction(
                currentState,
                secondAction,
                context,
              );
              currentState = secondOutcome.state;
              lastResult = secondOutcome.result;
              this.pushAction(actions, secondOutcome.result);

              const cardPlayed = p2Player.hand[p2CardIdx];
              if (this.hasSpecialEffect(cardPlayed)) {
                const effectOutcome = this.applyAction(
                  currentState,
                  applySpecialEffectAction(cardPlayed),
                  context,
                );
                currentState = effectOutcome.state;
                lastResult = effectOutcome.result;
                this.pushAction(actions, effectOutcome.result);
              }
            }
          }
        } catch (e) {
          this.logger.warn(`[AI][ONNX] Second prediction failed: ${String(e)}`);
        }
      }

      if (currentState.gameStatus !== 'finished') {
        const nextOutcome = this.applyAction(
          currentState,
          nextTurnAction(),
          context,
        );
        currentState = nextOutcome.state;
        lastResult = nextOutcome.result;
        this.pushAction(actions, nextOutcome.result);
      }

      return {
        state: currentState,
        done: currentState.gameStatus === 'finished',
        info: { aiActions: actions, source: 'onnx' },
      };
    } catch (error: unknown) {
      this.logger.error(
        `[AI][ONNX] fallback to rule-based due to: ${this.describeOnnxError(error)}`,
      );
      const fallback = this.executeTurn(state, context);
      return fallback
        ? {
            state: fallback.state,
            done: fallback.state.gameStatus === 'finished',
            info: {
              aiActions: fallback.actions,
              source: 'fallback',
              reason: this.describeOnnxError(error),
            },
          }
        : null;
    }
  }

  private describeOnnxError(error: unknown): string {
    if (
      error instanceof BadRequestException ||
      error instanceof NotFoundException
    ) {
      return `${error.name}: ${error.message}`;
    }
    if (error instanceof Error) {
      return `${error.name}: ${error.message}`;
    }
    return String(error);
  }

  public isAiTurn(state: GameState): boolean {
    const player = state.players.at(state.currentPlayerIndex);
    if (!player) {
      return false;
    }
    return player.isAI;
  }
}

File: onecard-server/src/onecard_api/services/game_ai_service.py
from __future__ import annotations

import json
import logging
from typing import Any

from onecard_api.domain.engine import (
    GameAction,
    apply_special_effect_action,
    draw_card_action,
    next_turn_action,
    play_card_action,
)
from onecard_api.domain.players import find_playable_card_brute_force
from onecard_api.domain.types import GameState, Player, PokerCard
from onecard_api.services.game_engine_service import GameEngineService
from onecard_api.services.onnx_policy_service import OnnxPolicyService

logger = logging.getLogger("onecard_api.game_ai")


class GameAiService:
    def __init__(
        self,
        game_engine: GameEngineService,
        onnx_policy_service: OnnxPolicyService,
    ) -> None:
        self._game_engine = game_engine
        self._onnx_policy_service = onnx_policy_service

    async def play_while_ai_turn(
        self, state: GameState, context: dict[str, Any] | None = None
    ) -> dict | None:
        if state["settings"]["mode"] != "single" or not self.is_ai_turn(state):
            return None

        if state["settings"]["difficulty"] == "medium":
            return await self._play_with_onnx(state, context)

        turn_result = self._execute_turn(state, context)
        if not turn_result:
            return None

        return {
            "state": turn_result["state"],
            "done": turn_result["state"]["gameStatus"] == "finished",
            "info": {"aiActions": turn_result["actions"]},
        }

    def _execute_turn(
        self, state: GameState, context: dict[str, Any] | None = None
    ) -> dict | None:
        player = state["players"][state["currentPlayerIndex"]]
        if not player.get("isAI"):
            return None

        current_state = state
        last_result: dict | None = None
        actions: list[GameAction] = []

        top_card = current_state["discardPile"][0] if current_state["discardPile"] else None
        card_to_play = (
            find_playable_card_brute_force(
                player["hand"], top_card, current_state["damage"]
            )
            if top_card
            else None
        )

        if card_to_play:
            playable_index = next(
                (idx for idx, c in enumerate(player["hand"]) if c["id"] == card_to_play["id"]),
                -1,
            )
            play_outcome = self._apply_action(
                current_state,
                play_card_action(current_state["currentPlayerIndex"], playable_index),
                context,
            )
            current_state = play_outcome["state"]
            last_result = play_outcome["result"]
            self._push_action(actions, play_outcome["result"])

            if self._has_special_effect(card_to_play):
                effect_outcome = self._apply_action(
                    current_state, apply_special_effect_action(card_to_play), context
                )
                current_state = effect_outcome["state"]
                last_result = effect_outcome["result"]
                self._push_action(actions, effect_outcome["result"])
        else:
            draw_outcome = self._apply_action(
                current_state,
                draw_card_action(max(1, current_state["damage"])),
                context,
            )
            current_state = draw_outcome["state"]
            last_result = draw_outcome["result"]
            self._push_action(actions, draw_outcome["result"])

        if current_state["gameStatus"] != "finished":
            next_outcome = self._apply_action(current_state, next_turn_action(), context)
            current_state = next_outcome["state"]
            last_result = next_outcome["result"]
            self._push_action(actions, next_outcome["result"])

        return {"state": current_state, "actions": actions, "result": last_result}

    def _apply_action(
        self,
        state: GameState,
        action: GameAction,
        context: dict[str, Any] | None = None,
    ) -> dict:
        self._log_ai_action(action, state["players"][state["currentPlayerIndex"]], context)
        result = self._game_engine.step(state, action)
        return {"state": result["state"], "result": result}

    def _push_action(self, actions: list[GameAction], result: dict) -> None:
        info_action = (result.get("info") or {}).get("action")
        if info_action:
            actions.append(info_action)  # type: ignore[arg-type]

    def _log_ai_action(
        self, action: GameAction, actor: Player | None, context: dict[str, Any] | None
    ) -> None:
        if not actor or not actor.get("isAI"):
            return
        metadata = {
            "event": "ai-action",
            "gameId": context.get("gameId") if context else None,
            "playerId": actor.get("id"),
            "playerName": actor.get("name"),
            "actionType": action.get("type"),
            "payload": action.get("payload") if isinstance(action, dict) else None,
        }
        try:
            serialized = json.dumps(metadata, ensure_ascii=False)
        except Exception:
            serialized = str(metadata)
        logger.info("[AI] %s", serialized)

    def _has_special_effect(self, card: PokerCard) -> bool:
        if card.get("isJoker"):
            return True
        special_ranks = {1, 2, 11, 12, 13}
        rank = card.get("rank")
        return rank in special_ranks if rank is not None else False

    async def _play_with_onnx(
        self, state: GameState, context: dict[str, Any] | None = None
    ) -> dict | None:
        try:
            current_state = state
            actions: list[GameAction] = []
            last_result: dict | None = None

            prediction = await self._onnx_policy_service.predict_action(current_state)
            payload = prediction["payload"]
            action_index = prediction["actionIndex"]
            is_draw = payload.get("type") == "DRAW_CARD"
            amount = payload.get("amount") or 1
            player_index = 0 if is_draw else payload.get("playerIndex", 0)
            card_index = 0 if is_draw else payload.get("cardIndex", 0)

            first_action: GameAction = (
                draw_card_action(int(amount))
                if is_draw
                else play_card_action(int(player_index), int(card_index))
            )

            logger.info(
                "[AI][ONNX] actionIndex=%s payload=%s",
                action_index,
                json.dumps(payload, ensure_ascii=False),
            )

            first_outcome = self._apply_action(current_state, first_action, context)
            current_state = first_outcome["state"]
            last_result = first_outcome["result"]
            self._push_action(actions, first_outcome["result"])

            if not is_draw:
                player = state["players"][state["currentPlayerIndex"]]
                card_to_play = (
                    player["hand"][card_index]
                    if 0 <= int(card_index) < len(player["hand"])
                    else None
                )
                if card_to_play and self._has_special_effect(card_to_play):
                    effect_outcome = self._apply_action(
                        current_state, apply_special_effect_action(card_to_play), context
                    )
                    current_state = effect_outcome["state"]
                    last_result = effect_outcome["result"]
                    self._push_action(actions, effect_outcome["result"])

            if current_state["gameStatus"] != "finished":
                next_outcome = self._apply_action(
                    current_state, next_turn_action(), context
                )
                current_state = next_outcome["state"]
                last_result = next_outcome["result"]
                self._push_action(actions, next_outcome["result"])

            return {
                "state": current_state,
                "done": current_state["gameStatus"] == "finished",
                "info": {"aiActions": actions, "source": "onnx"},
            }
        except Exception as error:  # broad catch to match JS fallback behaviour
            logger.error(
                "[AI][ONNX] fallback to rule-based due to: %s",
                self._describe_onnx_error(error),
            )
            fallback = self._execute_turn(state, context)
            if fallback:
                return {
                    "state": fallback["state"],
                    "done": fallback["state"]["gameStatus"] == "finished",
                    "info": {
                        "aiActions": fallback["actions"],
                        "source": "fallback",
                        "reason": self._describe_onnx_error(error),
                    },
                }
            return None

    def _describe_onnx_error(self, error: Exception) -> str:
        return f"{error.__class__.__name__}: {error}"

    def is_ai_turn(self, state: GameState) -> bool:
        player = state["players"][state["currentPlayerIndex"]]
        return bool(player.get("isAI"))
